// angular-epub-viewer/esm5/angular-epub-viewer.js
import { Component, ElementRef, EventEmitter, Input, NgZone, Output, ViewChild, ViewEncapsulation, NgModule } from '@angular/core';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import 'rxjs/add/operator/filter';

var EpubError = {
    OPEN_FILE: 0,
    READ_FILE: 1,
    NOT_LOADED_DOCUMENT: 2,
    NOT_DISPLAYED_CHAPTER: 3,
    SEARCH: 4,
    COMPUTE_PAGINATION: 5,
    LOAD_METADATA: 6,
    LOAD_TOC: 7,
};
EpubError[EpubError.OPEN_FILE] = "OPEN_FILE";
EpubError[EpubError.READ_FILE] = "READ_FILE";
EpubError[EpubError.NOT_LOADED_DOCUMENT] = "NOT_LOADED_DOCUMENT";
EpubError[EpubError.NOT_DISPLAYED_CHAPTER] = "NOT_DISPLAYED_CHAPTER";
EpubError[EpubError.SEARCH] = "SEARCH";
EpubError[EpubError.COMPUTE_PAGINATION] = "COMPUTE_PAGINATION";
EpubError[EpubError.LOAD_METADATA] = "LOAD_METADATA";
EpubError[EpubError.LOAD_TOC] = "LOAD_TOC";
var AngularEpubViewerComponent = /** @class */ (function () {
    function AngularEpubViewerComponent(zone) {
        var _this = this;
        this.zone = zone;
        this.epub = null;
        this.currentLocation = {
            startCfi: null,
            endCfi: null,
            page: null,
            chapter: null
        };
        this.documentReady = false;
        this.isChapterDisplayed = false;
        this.computingPagination = false;
        this.searchingText = false;
        this.padding = null;
        this.autoPagination = false;
        this.autoMetadata = false;
        this.autoTOC = false;
        this.onDocumentReady = new EventEmitter();
        this.onChapterUnloaded = new EventEmitter();
        this.onChapterDisplayed = new EventEmitter();
        this.onLocationFound = new EventEmitter();
        this.onSearchFinished = new EventEmitter();
        this.onPaginationComputed = new EventEmitter();
        this.onMetadataLoaded = new EventEmitter();
        this.onTOCLoaded = new EventEmitter();
        this.onErrorOccurred = new EventEmitter();
        this._link = new BehaviorSubject(null);
        this.needSearchText = null;
        this.needComputePagination = false;
        this.initEpub = function (properties) {
            _this.destroyEpub();
            _this.epub = ePub(properties);
            _this.epub.on('book:ready', function () {
                _this.zone.run(function () {
                    _this.documentReady = true;
                    _this.onDocumentReady.next(null);
                    if (_this.autoPagination) {
                        _this.needComputePagination = true;
                    }
                    if (_this.autoMetadata) {
                        _this.loadMetadata();
                    }
                    if (_this.autoTOC) {
                        _this.loadTOC();
                    }
                });
            });
            _this.epub.on('book:pageChanged', function (location) {
                _this.zone.run(function () {
                    if (!_this.computingPagination) {
                        _this.currentLocation.page = location.anchorPage;
                        _this.onLocationFound.next(_this.currentLocation);
                    }
                });
            });
            _this.epub.on('renderer:chapterUnloaded', function () {
                _this.zone.run(function () {
                    _this.isChapterDisplayed = false;
                    _this.onChapterUnloaded.next(null);
                });
            });
            _this.epub.on('renderer:chapterDisplayed', function (chapter) {
                _this.zone.run(function () {
                    _this.isChapterDisplayed = true;
                    chapter['label'] = null;
                    _this.onChapterDisplayed.next(chapter);
                    _this.currentLocation.chapter = chapter;
                    _this.onLocationFound.next(_this.currentLocation);
                    if (_this.needComputePagination) {
                        _this.computePagination();
                    }
                });
            });
            _this.epub.on('renderer:resized', function () {
                _this.zone.run(function () {
                    _this.needComputePagination = true;
                    if (_this.autoPagination) {
                        _this.computePagination();
                    }
                });
            });
            _this.epub.on('renderer:visibleRangeChanged', function (range) {
                _this.zone.run(function () {
                    _this.currentLocation.startCfi = range.start;
                    _this.currentLocation.endCfi = range.end;
                    _this.onLocationFound.next(_this.currentLocation);
                });
            });
            _this.epub.renderTo('angularEpubViewerComponent');
        };
    }
    AngularEpubViewerComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.linkSubscription = this._link.asObservable()
            .filter(function (link) { return link != null; })
            .subscribe(function (link) {
            _this.initEpub({
                bookPath: link
            });
        });
    };
    AngularEpubViewerComponent.prototype.openLink = function (link) {
        this._link.next(link);
    };
    AngularEpubViewerComponent.prototype.openFile = function (file) {
        var _this = this;
        if (window['FileReader']) {
            this.zone.runOutsideAngular(function () {
                var reader = new FileReader();
                reader.onload = function () {
                    _this.zone.run(function () {
                        _this.initEpub({
                            bookPath: reader.result
                        });
                    });
                };
                reader.onerror = function () {
                    _this.zone.run(function () {
                        _this.onErrorOccurred.emit(EpubError.READ_FILE);
                    });
                };
                reader.readAsArrayBuffer(file);
            });
        }
        else {
            this.onErrorOccurred.emit(EpubError.OPEN_FILE);
        }
    };
    AngularEpubViewerComponent.prototype.goTo = function (location) {
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        if (typeof location === "number") {
            this.epub.displayChapter(location);
        }
        else if (/.*html$/.test(location)) {
            this.epub.goto(location);
        }
        else {
            this.epub.displayChapter(location);
        }
    };
    AngularEpubViewerComponent.prototype.nextPage = function () {
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        this.epub.nextPage();
    };
    AngularEpubViewerComponent.prototype.previousPage = function () {
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        this.epub.prevPage();
    };
    AngularEpubViewerComponent.prototype.searchText = function (text) {
        var _this = this;
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        if (!this.isChapterDisplayed) {
            this.onErrorOccurred.emit(EpubError.NOT_DISPLAYED_CHAPTER);
            return;
        }
        if (!text || text.trim().length <= 0) {
            this.onErrorOccurred.emit(EpubError.SEARCH);
            return;
        }
        if (this.searchingText) {
            this.needSearchText = text;
            return;
        }
        this.searchingText = true;
        this.needSearchText = null;
        this.zone.runOutsideAngular(function () {
            var results = _this.epub.currentChapter.find(text.trim());
            _this.zone.run(function () {
                _this.searchingText = false;
                if (_this.needSearchText) {
                    _this.searchText(_this.needSearchText);
                }
                else {
                    _this.onSearchFinished.next(results);
                }
            });
        });
    };
    AngularEpubViewerComponent.prototype.setStyle = function (style, value) {
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        this.epub.setStyle(style, value);
    };
    AngularEpubViewerComponent.prototype.resetStyle = function (style) {
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        this.epub.removeStyle(style);
    };
    AngularEpubViewerComponent.prototype.computePagination = function () {
        var _this = this;
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        if (!this.isChapterDisplayed) {
            this.onErrorOccurred.emit(EpubError.NOT_DISPLAYED_CHAPTER);
            return;
        }
        if (this.computingPagination) {
            return;
        }
        this.computingPagination = true;
        this.needComputePagination = false;
        this.zone.runOutsideAngular(function () {
            _this.epub.generatePagination()
                .then(function (pages) {
                var currentPage = _this.epub.pagination.pageFromCfi(_this.epub.getCurrentLocationCfi());
                _this.zone.run(function () {
                    _this.computingPagination = false;
                    if (_this.needComputePagination) {
                        _this.computePagination();
                    }
                    else {
                        _this.onPaginationComputed.next(pages);
                        _this.currentLocation.page = currentPage;
                        _this.onLocationFound.next(_this.currentLocation);
                    }
                });
            })
                .catch(function () {
                _this.zone.run(function () {
                    _this.computingPagination = false;
                    _this.onErrorOccurred.emit(EpubError.COMPUTE_PAGINATION);
                });
            });
        });
    };
    AngularEpubViewerComponent.prototype.loadMetadata = function () {
        var _this = this;
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        this.zone.runOutsideAngular(function () {
            _this.epub.getMetadata()
                .then(function (metadata) {
                _this.zone.run(function () {
                    _this.onMetadataLoaded.next(metadata);
                });
            })
                .catch(function () {
                _this.zone.run(function () {
                    _this.onErrorOccurred.emit(EpubError.LOAD_METADATA);
                });
            });
        });
    };
    AngularEpubViewerComponent.prototype.loadTOC = function () {
        var _this = this;
        if (!this.documentReady) {
            this.onErrorOccurred.emit(EpubError.NOT_LOADED_DOCUMENT);
            return;
        }
        this.zone.runOutsideAngular(function () {
            _this.epub.getToc()
                .then(function (chapters) {
                _this.zone.run(function () {
                    _this.onTOCLoaded.next(chapters);
                });
            })
                .catch(function () {
                _this.zone.run(function () {
                    _this.onErrorOccurred.emit(EpubError.LOAD_TOC);
                });
            });
        });
    };
    AngularEpubViewerComponent.prototype.destroyEpub = function () {
        this.documentReady = false;
        this.isChapterDisplayed = false;
        this.searchingText = false;
        this.needSearchText = null;
        this.computingPagination = false;
        this.needComputePagination = false;
        this.currentLocation = {
            startCfi: null,
            endCfi: null,
            page: null,
            chapter: null
        };
        if (this.epub) {
            this.epub.destroy();
            this.epub = null;
        }
    };
    
    AngularEpubViewerComponent.prototype.ngOnDestroy = function () {
        if (this.linkSubscription) {
            this.linkSubscription.unsubscribe();
        }
        this.destroyEpub();
    };
    return AngularEpubViewerComponent;
}());
AngularEpubViewerComponent.decorators = [
    { type: Component, args: [{
                selector: 'angular-epub-viewer',
                template: "<div id=\"angularEpubViewerComponent\" [style.padding]=\"padding\" #angularEpubViewerComponent></div>",
                styles: ["\n        #angularEpubViewerComponent {\n            width: 100%;\n            height: 100%;\n            margin: 0;\n            overflow: hidden;\n        }\n    "],
                encapsulation: ViewEncapsulation.None
            },] },
];
AngularEpubViewerComponent.ctorParameters = function () { return [
    { type: NgZone, },
]; };
AngularEpubViewerComponent.propDecorators = {
    "root": [{ type: ViewChild, args: ['angularEpubViewerComponent', { read: ElementRef },] },],
    "padding": [{ type: Input },],
    "autoPagination": [{ type: Input },],
    "autoMetadata": [{ type: Input },],
    "autoTOC": [{ type: Input },],
    "onDocumentReady": [{ type: Output, args: ['onDocumentReady',] },],
    "onChapterUnloaded": [{ type: Output, args: ['onChapterUnloaded',] },],
    "onChapterDisplayed": [{ type: Output, args: ['onChapterDisplayed',] },],
    "onLocationFound": [{ type: Output, args: ['onLocationFound',] },],
    "onSearchFinished": [{ type: Output, args: ['onSearchFinished',] },],
    "onPaginationComputed": [{ type: Output, args: ['onPaginationComputed',] },],
    "onMetadataLoaded": [{ type: Output, args: ['onMetadataLoaded',] },],
    "onTOCLoaded": [{ type: Output, args: ['onTOCLoaded',] },],
    "onErrorOccurred": [{ type: Output, args: ['onErrorOccurred',] },],
};
var AngularEpubViewerModule = /** @class */ (function () {
    function AngularEpubViewerModule() {
    }
    return AngularEpubViewerModule;
}());
AngularEpubViewerModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [AngularEpubViewerComponent],
                exports: [AngularEpubViewerComponent]
            },] },
];

export { AngularEpubViewerModule, AngularEpubViewerComponent, EpubError };
